lib/incoming_mail/handler.rb

```diff
@@ -63,19 +63,25 @@ def create_comment(msg)
       proposal = find_proposal(find_public_id(msg)) or return
       comment_text = find_comment_text(msg)
       comment_user = find_comment_user(msg)
-      comment = nil
-      if proposal.has_subscriber?(comment_user)
-        # already in the loop, just add comment.
-        comment = proposal.comments.create(comment_text: comment_text, user: comment_user)
-      else
-        # yes, user adds self as observer, which also generates comment
-        proposal.add_observer(comment_user, comment_user, comment_text)
-        comment = proposal.comments.last
+      # IMPORTANT that we check/add as observer before we create comment,
+      # since comment will create as a user if not already,
+      # and we want the reason logged.
+      unless proposal.has_subscriber?(comment_user)
+        add_user_as_observer(comment_user, proposal)
       end
+      comment = proposal.comments.create(comment_text: comment_text, user: comment_user)
       Dispatcher.on_comment_created(comment) # sends email
       comment
     end
 
+    def add_user_as_observer(user, proposal)
+      observer_role = Role.find_or_create_by(name: 'observer')
+      observation = Observation.new(user_id: user.id, role_id: observer_role.id, proposal_id: proposal.id)
+      proposal.observations << observation
+      Dispatcher.on_observer_added(observation, "Added comment via email reply")
+      observation
+    end
```

Comment:
It does, mostly. The key difference is that it does not create a comment with the `update_comment` flag set, which has the effect of making `Dispatcher.on_observer_added` a no-op (no email is sent).

I thought of refactoring `Proposal.create_new_observation` and/or adding a new method to `Proposal` but that felt pretty intrusive and I wasn't convinced would actually end up saving any lines of code.

I'm happy to see this refactored eventually. In the meantime, I am going to merge and deploy since it fixes the immediate problem.